src/dynimage.rs

```diff
@@ -910,6 +928,15 @@ pub fn load<R: BufRead + Seek>(r: R, format: ImageFormat) -> ImageResult<Dynamic
     }
 }
 
+fn load_guess<R: BufRead + Seek>(mut r: R) -> ImageResult<DynamicImage> {
+    let mut buffer: [u8; 16] = [0u8; 16];
+    r.read(&mut buffer[..])?;
```

Comment:
This should probably use `read_exact` so that the loading doesn't fail if `read` only reads a few bytes.

Reply:
When I wrote that I was considering what would happen if input file is < 16 bytes, then .read() should fill buffer with available bytes.
Since buffer is 16 bytes big, I think read_exact() will error then.
But I'm just speculating about the behavior really, didn't test any of it.

Reply:
Pushed change to use read_exact

Reply:
I am pretty sure that `read_exact` will error, but you would probably get a similar error from all of the decoders. I guess no image file can be less than 16 bytes long?

Reply:
Yea I don't think I a valid image file could be <16 bytes.

Reply:
You can comfortably fit a pnm in less than that space even with an unecessary comment on its content: (15 bytes)

```
P1\n1 1#white\n\n1
```

Reply:
@HeroicKatora What do you think, should we revert this behaviour then in order to handle a 15 byte PBM ?

Reply:
You could simply shorten some of the magic byte constants so that none of them are more than 5 bytes. They should still be uniquely identifiable I think? If this turns out to be a problem later, we should probably consider a slightly smarter system.

Reply:
An other solution is to use something like and simply fill as much as possible under 16 bytes:
```rust
let mut buf = [0; 16];
let mut len = 0;
while !buf.is_empty() {
    match self.read(&mut buf[len..]) {
        Ok(0) => break,
        Ok(n) => len += n,
        Err(ref e) if e.kind() == ErrorKind::Interrupted => (),
        Err(e) => return Err(e),
    }
}
```

Reply:
Probably even better would be to use:
```rust
match r.read_exact(&mut buf) {
    Ok(()) => (),
    Err(e) if e.kind() == ErrorKind::UnexpectedEof => (),
    Err(e) => return Err(e.into()),
}
```

Reply:
No, don't revert it. There are more options than those mentioned by @birktj . You could wrap the buffer into a `io::Cursor` to read potentially less bytes and use `take` to limit the maximum number:

```
    let mut buffer: [u8; 16] = [0u8; 16];
    let mut out = io::Cursor::new(&mut buffer[..]);
    let cur = r.seek(SeekFrom::Current(0))?;
    let len = io::copy(&mut cur.take(16), &mut out)?;
    r.seek(SeekFrom::Start(cur))?;
    load(r, guess_format(&buffer[..len as usize])?)
```

Reply:
No live updates on review conversations so now you have ample alternatives @martinlindhe

Reply:
I'll hopefully get back to this shortly- if you are reading this from the future and I didn't, please ping me!